### ISSUE TITLE
TensorRT converter fails for CombinedNonMaxSuppression

### ISSUE THREAD

**Issue by srihari-humbarwadi** [tensorflow/tensorflow#46453]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Ubuntu 18.04**
- TensorFlow installed from (source or binary): **Binary**
- TensorFlow version (use command below): **TF:2.5.0-dev20210114**
- Python version: **3.7**
- CUDA/cuDNN version: **11.0, 8.0.4**
- GPU model and memory: **1060**

**Describe the current behavior**
TensorRT converter crashes with a segmentation fault when I try to export my `saved_model`.
Interestingly, if I set `minimum_segment_size=10`, it works because it skips 

*Replaced segment 5 consisting of 7 nodes by StatefulPartitionedCall/decode_predictions/TRTEngineOp_0_5.
2021-01-15 15:21:38.915310: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:858] Segment consists of nodes: StatefulPartitionedCall/decode_predictions/combined_non_max_suppression/CombinedNonMaxSuppression, StatefulPartitionedCall/decode_predictions/combined_non_max_suppression/CombinedNonMaxSuppression/max_output_size_per_class, StatefulPartitionedCall/decode_predictions/combined_non_max_suppression/Const, StatefulPartitionedCall/decode_predictions/combined_non_max_suppression/iou_threshold, StatefulPartitionedCall/decode_predictions/combined_non_max_suppression/score_threshold, StatefulPartitionedCall/decode_predictions/transpose_1, StatefulPartitionedCall/decode_predictions/transpose_1/perm*

I have attached the full log after running with these flags
`TF_CPP_VMODULE=trt_engine_op=2,convert_nodes=2,convert_graph=2,segment=2,trt_shape_optimization_profiles=2,trt_engine_resource_ops=2 python trt.py`

**Standalone code to reproduce the issue**
```python
import os

import tensorflow as tf

## Download and extract the zip 
## URL: https://drive.google.com/file/d/1Zxqdnm2iHpJGdUl17cAi-lV7wZ3UhMDA/view

params = tf.experimental.tensorrt.ConversionParams(
    precision_mode='FP32',
    maximum_cached_engines=1,
    minimum_segment_size=5)

converter = tf.experimental.tensorrt.Converter(
    input_saved_model_dir='retinanet-18-640-30x-64-tpu',
    conversion_params=params)
converter.convert()

def input_fn(steps=1):
    for i in range(steps):
        yield (tf.random.uniform([640, 640, 3]), tf.constant(1, dtype=tf.int32))
        
converter.build(input_fn=input_fn)
converter.save('trt')
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
[trt_log.txt](https://github.com/tensorflow/tensorflow/files/5819748/trt_log.txt)

**Comment by amahendrakar** [comment]
Was able to run the code without any issues on [TF v2.4](https://colab.research.google.com/gist/amahendrakar/7f400de432bddbf6b4e47a0feb33ed7a/46453.ipynb).

However, session crashes on running the code with [TF-nightly](https://colab.research.google.com/gist/amahendrakar/be033a6c68b0028f5522924ef378e66b/46453-tf-nightly.ipynb#scrollTo=rdbO5RSdFM6a) (i.e. v2.5.0-dev20210114). Please check the linked gist for reference. Thanks!

**Comment by srihari-humbarwadi** [comment]
I can successfully convert, build and then load back the model in `2.4.0` but the same code fails in nightly `dev20210114`

**Comment by sanjoy** [comment]
@bixia1 what do you think?  Is [this](https://github.com/tensorflow/tensorflow/blob/4fa4184a5a454eceb5b567c8b3c4fce46faf2de8/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L5914) correct?

**Comment by bixia1** [comment]
The [PR in question](https://github.com/tensorflow/tensorflow/pull/40062) (merged on 1/13/2021) doesn’t make any change in terms of always use normalized coordinate in TF-TRT.

The [TF document](https://www.tensorflow.org/api_docs/python/tf/image/combined_non_max_suppression) says the bounding box coordinates can be normalized or absolute values. But I couldn't tell how the operation representation indicate whether the coordinates are normalized or absolute values.  I looked at [the implement of the compute method for the operation](https://github.com/tensorflow/tensorflow/blob/9045dcaf276cb7b24fde33da09165cb38d157a5e/tensorflow/core/kernels/image/non_max_suppression_op.cc#L916) and couldn't figure out this either. I am asking for information about this in an internal channel.

**Comment by bixia1** [comment]
@tfeher

**Comment by bixia1** [comment]
I got an answer from the TensorFlow people which help me understand the situation. The TensorFlow implementation for the operation is the same regardless whether the coordinates are normalized or absolute values. On the other hand, the TensorRT implementation requires a parameter that specifies whether the coordinates are normalized or not, see code [here](https://github.com/tensorflow/tensorflow/blob/ab9eb3d10400c92b15bfae62b284bf9937cb1845/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L5951-L5952). If this field is indeed used in the TensorRT implementation, then we have a problem here. Now the question is why it was working before the  [PR in question](https://github.com/tensorflow/tensorflow/pull/40062)? I will let @tfeher and @DEKHTIARJonathan take care of this.

**Comment by srihari-humbarwadi** [comment]
@bixia1 There are two issues
 - Conversion of `tf.image.combined_non_max_suppression` in nightly fails. But works in `2.4.0`
 - Reading the comments [here](https://github.com/tensorflow/tensorflow/blob/4fa4184a5a454eceb5b567c8b3c4fce46faf2de8/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L5911-L5917), calculation of width/height differs in tensorrt vs tensorflow op. To avoid this the converter always assumes the users are sending normalized coordinates [here](https://github.com/tensorflow/tensorflow/blob/4fa4184a5a454eceb5b567c8b3c4fce46faf2de8/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L5918)

Hence, if the users want to export `tf.image.combined_non_max_suppression` to tensorrt, they should only send normalized coordinates. But the documentation fails to warn the user about this.

**Comment by tfeher** [comment]
> The TensorFlow implementation for the operation is the same regardless whether the coordinates are normalized or absolute values. On the other hand, the TensorRT implementation requires a parameter that specifies whether the coordinates are normalized or not

TensorRT indeed has two implementation for the IOU calculation one for coordinates that should be interpreted as [pixels](https://github.com/NVIDIA/TensorRT/blob/183f891191f08fd016216fd0b94bc9c8c52d0ac2/plugin/common/kernels/allClassNMS.cu#L142-L143), and one otherwise (activated by[ isNormalized=true](https://github.com/NVIDIA/TensorRT/blob/183f891191f08fd016216fd0b94bc9c8c52d0ac2/plugin/common/kernels/allClassNMS.cu#L135-L139)). Note that `isNormalized` is an unfortunate name for the option, what it does is simply switching between these two modes. The latter implementation agrees with the [TF implementation](https://github.com/tensorflow/tensorflow/blob/508374a893df7999633d9ebbe55e94d92eef7280/tensorflow/core/kernels/image/non_max_suppression_op.cc#L123-L134), therefore the converter uses only that mode [mode of the TRT plugin](https://github.com/tensorflow/tensorflow/blob/ac4b2997b6c47e06a4f689355c08418856905594/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L5911-L5918). We do have an unit test which checks that the converter works for non-normalized coordinates. 

To summarize, the TF-TRT converted combinedNMS op should work with not-normalized coordinates. #40062 did not touch this. Among other things, the handling of the `clib_boxes` attribute was corrected in that PR, but not sure if that is relevant here.

I will have a closer look at the problem and report back.

**Comment by tfeher** [comment]
I could reproduce the bug. The TRT engine throws a segfault when we call [enqueue](https://github.com/tensorflow/tensorflow/blob/3db793ee031f2abede180043b016c085f1d8c26b/tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc#L280). The problem indeed happens after the converter was changed in #40062, if I revert the changes then the segfault disappears. It still needs to be clarified whether we make an error on TF-TRT side while setting up the TRT NMS plugin parameters, or it is a bug in TRT.

**Comment by tfeher** [comment]
The issue is caused by the change in handling the `top_k` parameter for the plugin. The value that we are passing (5000) is larger than what TRT can handle (4096). There are two options to fix this:
- Cap top_k to 4096. We have to check whether this leads to incompatibility between the TF and TRT results.
- Mark node as incompatible, this is always safe but bad for performance.

**Comment by srihari-humbarwadi** [comment]
> > The TensorFlow implementation for the operation is the same regardless whether the coordinates are normalized or absolute values. On the other hand, the TensorRT implementation requires a parameter that specifies whether the coordinates are normalized or not
> 
> TensorRT indeed has two implementation for the IOU calculation one for coordinates that should be interpreted as [pixels](https://github.com/NVIDIA/TensorRT/blob/183f891191f08fd016216fd0b94bc9c8c52d0ac2/plugin/common/kernels/allClassNMS.cu#L142-L143), and one otherwise (activated by[ isNormalized=true](https://github.com/NVIDIA/TensorRT/blob/183f891191f08fd016216fd0b94bc9c8c52d0ac2/plugin/common/kernels/allClassNMS.cu#L135-L139)). Note that `isNormalized` is an unfortunate name for the option, what it does is simply switching between these two modes. The latter implementation agrees with the [TF implementation](https://github.com/tensorflow/tensorflow/blob/508374a893df7999633d9ebbe55e94d92eef7280/tensorflow/core/kernels/image/non_max_suppression_op.cc#L123-L134), therefore the converter uses only that mode [mode of the TRT plugin](https://github.com/tensorflow/tensorflow/blob/ac4b2997b6c47e06a4f689355c08418856905594/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L5911-L5918). We do have an unit test which checks that the converter works for non-normalized coordinates.
> 
> To summarize, the TF-TRT converted combinedNMS op should work with not-normalized coordinates. #40062 did not touch this. Among other things, the handling of the `clib_boxes` attribute was corrected in that PR, but not sure if that is relevant here.
> 
> I will have a closer look at the problem and report back.

@tfeher 
https://github.com/tensorflow/tensorflow/blob/a6f927400dfa09f445faee953e5694512226eed5/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L5980-L5987
Can you please clarify: Since `is_normalized` explicitly being set to `true`, does this mean that the converter expects the coordinates to be normalized `[0, 1]`?. When I tried passing unnormalized coordinates, I did not get correct results.

**Comment by tfeher** [comment]
As I have described above, the combinedNMS op is expected to work with not normalized coordinates. And yes, we need to set `is_normalized = true` even if we have not normalied coordinates. TRT's naming of the argument is somewhat unfortunate. I have confirmed with a TRT engineer that the `isNormalized` arg only switches modes of IOU calculation, it does not require us to provide normalized coordinates.

Note that TRT has an option `clipBoxes` with the following meaning:

> Forcibly restrict bounding boxes to the normalized range [0,1]. Only applicable if isNormalized is also true. Defaults to true.

If it is set to false, then the conversion should work with unnormalized coordinates.

Now the problem is, that before #40062, the `clipBoxes` arg was not set by the converter, so it defaulted to truncating the box coordinates. This is one of the things which was fixed by #40062. 

The question is, which version of TF are you using  when you get the incorrect results?

**Comment by srihari-humbarwadi** [comment]
@tfeher, I guess this explains the behaviour. I was successfully able to convert the models in `2.4.0` but the output boxes were being clipped to 1 for the converted model. 
> clipBoxes arg was not set by the converter

**Comment by bixia1** [comment]
@tfeher, we encountered similar crash for a google internal customer and verified that setting top_k to 4096 fix the problem. For their case, num_boxes is 70K+. 

I think we probably misunderstand the relationship between num_boxes and top_k in [this code block](https://github.com/tensorflow/tensorflow/blob/ed22f400428a669c1c6e4553cd7f4900abeaf954/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L5999-L6002).  top_k in the plugin is more like the num of top values we keep in the internal of the algorithm, in order to support the selection of keep_top_k as output, is it?

I think we should fix this to something like this:
if (keep_top_k > 4096) return "tensorrt not support" else top_k = 4096 (or top_k = keep_top_k?)

They also help me check the performance for these two ways of setting top_k:
     top_k = keep_top_k
     top_k = 4096
and didn't see any perf diff for their app.

**Comment by bixia1** [comment]
See cloud_tpu is setting top_k to 5000 [here](https://github.com/tensorflow/tpu/blob/d4daff70a8a17625cb43386b2a564cb0e0e0e130/models/official/detection/ops/postprocess_ops.py#L60-L80)
Here is their definition of top_k: 
pre_nms_num_boxes: an int number of top candidate detections per class
      before NMS

**Comment by srihari-humbarwadi** [comment]
@bixia1 pre_nms_top_k is used in models like RetinaNet, EfficientDet and other similar models. And in most of the cases, the literature asks us to pick the top 5000 boxes (depending on the score), this is where the number 5000 comes from.

But what troubles me is there is no way in which the user can set this value when calling `combined_non_max_suppression`. I feel that since tensorrt plugin already has this field, it would be beneficial to expose this param in `tf.image.combined_non_max_suppression`

```python
tf.image.combined_non_max_suppression(
    boxes, scores, max_output_size_per_class, max_total_size, iou_threshold=0.5,
    score_threshold=float('-inf'), pad_per_class=False, clip_boxes=True,
    name=None
)
```

**Comment by bixia1** [comment]
Hi, Tama's PR is going to fix this by introducing an environment for you to overwrite the default behavior. With his PR, the default behavior won't be changed, that is, we still reject the case where top_k > 4096 as we do currently unless you explicit request to change this behavior through providing the environment variable. We need your feedback on this solution.
